cilium-dbg/cmd/bpf_multicast_subscribers.go

```diff
@@ -0,0 +1,294 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package cmd
+
+import (
+	"errors"
+	"fmt"
+	"io/fs"
+	"net/netip"
+	"os"
+	"sort"
+	"text/tabwriter"
+
+	"github.com/spf13/cobra"
+	"github.com/vishvananda/netlink"
+
+	"github.com/cilium/cilium/pkg/command"
+	"github.com/cilium/cilium/pkg/common"
+	"github.com/cilium/cilium/pkg/defaults"
+	maps_multicast "github.com/cilium/cilium/pkg/maps/multicast"
+)
+
+const (
+	allGroupsKW = "all"
+	remoteKW    = "Remote Node"
+	localKW     = "Local Endpoint"
+)
+
+var BpfMcastSubscriberCmd = &cobra.Command{
+	Use:     "subscriber",
+	Aliases: []string{"sub"},
+	Short:   "Manage the multicast subscribers.",
+}
+
+func init() {
+	BpfMcastSubscriberCmd.AddCommand(MulticastGroupSubscriberListCmd)
+	command.AddOutputOption(MulticastGroupSubscriberListCmd)
+	BpfMcastSubscriberCmd.AddCommand(MulticastGroupSubscriberAddCmd)
+	BpfMcastSubscriberCmd.AddCommand(MulticastGroupSubscriberDelCmd)
+}
+
+var MulticastGroupSubscriberListCmd = &cobra.Command{
+	Use:     "list < group | all >",
+	Aliases: []string{"ls"},
+	Short:   "List the multicast subscribers for the given group.",
+	Long: `List the multicast subscribers for the given group. 
+	To get the subscribers for all the groups, use 'cilium-dbg bpf multicast subscriber list all'.`,
+	Run: func(cmd *cobra.Command, args []string) {
+		common.RequireRootPrivilege("cilium bpf multicast subscriber list")
+
+		groupAddr, all, err := parseMulticastGroupSubscriberListArgs(args)
+		if err != nil {
+			Fatalf("invalid argument: %s\n", err)
+		}
+
+		groupV4Map, err := maps_multicast.OpenGroupV4OuterMap(maps_multicast.GroupOuter4MapName)
+		if err != nil {
+			if errors.Is(err, fs.ErrNotExist) {
+				fmt.Fprintln(os.Stderr, "Cannot find multicast bpf maps")
+				return
+			}
+			Fatalf("Failed to open multicast bpf maps: %s", err)
+		}
+
+		var groups []netip.Addr
+		if all {
+			groups, err = groupV4Map.List()
+			if err != nil {
+				Fatalf("Failed to list multicast groups: %s", err)
+			}
+		} else {
+			groups = append(groups, groupAddr)
+		}
+
+		err = outputGroups(groupV4Map, groups)
```

Comment:
We could dry this up. outputGroups has multiple concerns. validating the ip and printing. In the non `all` case we could just check the group ip here prior to adding it to the slice and outputGroups now just prints since in the `all` case the ips would be valid?

Reply:
Hmm, not sure if I follow this as well. Yes, outputGroups is doing two things - getting subscriber data for each group and printing this information. I didn't really had another consumer of subscriber data so kept the two things in same method. 

Even in case of single group, we need to get subscriber list of that group - which will be slice of `maps_multicast.SubscriberV4`.

Reply:
Hi @derailed let me know if current code here makes sense ? My thinking here is to treat both cases ( all, single group ) in same way - to avoid duplicating code for getting subscriber list.